Add support to fetch a specific group of status json fields

sbodagala opened the pull request:
Add a special key, in METRICS module, in order to fetch a specific group of "status json" fields from FoundationDB.

Testing:

- Added a test in SpecialKeySpaceCorrectness (the test succeeds).
- Joshua job: 20231006-161737-sre-5681dc51841bc8e2 (in progress).

# Code-Reviewer Section

The general pull request guidelines can be found [here](https://github.com/apple/foundationdb/wiki/FoundationDB-Commit-Process).

Please check each of the following things and check *all* boxes before accepting a PR.

- [ ] The PR has a description, explaining both the problem and the solution.
- [ ] The description mentions which forms of testing were done and the testing seems reasonable.
- [ ] Every function/class/actor that was touched is reasonably well documented.

## For Release-Branches

If this PR is made against a release-branch, please also check the following:

- [ ] This change/bugfix is a cherry-pick from the next younger branch (younger `release-branch` or `main` if this is the youngest branch)
- [ ] There is a good reason why this PR needs to go into a release branch and this reason is documented (either in the description above or in a linked GitHub issue)

sbodagala commented on fdbserver/Status.actor.cpp:
```diff
@@ -3032,7 +3032,13 @@ ACTOR Future<StatusReply> clusterGetStatus(
+    metacluster::MetaclusterMetrics metaclusterMetrics,
+    std::string requestedStatusJsonField) {
+
+	if (requestedStatusJsonField == "fault_tolerance") {
```
There is a lot of code duplication here, so I am not very excited about making this actor.

Other approach is to get the entire status json, and then return only the requested fields out of this. But this would make the actor do a lot of unnecessary work in that case.

jzhou77 replied:
To refactor the code, we might need something like a mapping table: `status_field -> function`, such that arbitrary fields can be satisfied by calling several functions. I'm open to other ideas as well.

sbodagala replied:
I see what you are saying but the thing is there are a number of status fields (their count is quite high) and also there is a dependency among these (based on how they are populated/the code is organized).

jzhou77 commented on fdbserver/ClusterController.actor.cpp:
```diff
@@ -1489,54 +1489,60 @@ ACTOR Future<Void> statusServer(FutureStream<StatusRequest> requests,
-			std::vector<WorkerDetails> workers;
-			std::vector<ProcessIssues> workerIssues;
+			while (!requests_batch.empty()) {
+				state std::string requestedStatusJsonField = requests_batch.back().statusField;
```
Another idea is to accumulate all the `statusField` in the `requests_batch`, and only issue one `clusterGetStatus` call. After the call returns, find values for each `statusField` and send them back.

sbodagala replied:
Yes, I like this suggestion. This kind of resembles what I said in my comment:

> Other approach is to get the entire status json, and then return only the requested fields out of this. But this would make the actor do a lot of unnecessary work in that case.

The benefits of this approach are that:

- we wouldn't need to modify "clusterGetStatus()" (and/or add actors that duplicate code)
- invoke "clusterGetStatus()" once per batch

Disadvantages being:

- we will do the job of populating the entire status (and making the network calls between FoundationDB processes in order to populate the status) even if the caller/batch requests a single field (or a small subset of the entire status)

- we will need to do the additional task of fetching and populating the requested fields from the complete status json output string (but likely this isn't going to add a lot of overhead).

sbodagala replied:
Modified code to follow this approach. Please review, thanks!

jzhou77 commented on fdbserver/ClusterController.actor.cpp:
```diff
@@ -1489,54 +1489,60 @@ ACTOR Future<Void> statusServer(FutureStream<StatusRequest> requests,
+				// requests
+				last_request_time = now();
+
+				while (!requests_batch.empty() && requests_batch.back().statusField == requestedStatusJsonField) {
```
it's possible that the last one in the batch doesn't match `statusField`, and the loop exits. However, there can be requests before that match the `statusField`, resulting in unnecessary fetches.

sbodagala replied:
The "statusField" of the last entry has to match "requestedStatusJsonField", right? This is because "requestedStatusJsonField" is populated based on the last entry:

`state std::string requestedStatusJsonField = requests_batch.back().statusField;`

jzhou77 commented on fdbserver/Status.actor.cpp:
```diff
@@ -3669,6 +3670,45 @@ ACTOR Future<StatusReply> clusterGetStatus(
 		    .detail("Duration", timer() - tStart)
 		    .detail("StatusSize", statusObj.getFinalLength());
 
+		std::cout << "StatusObj string: " << statusObj.getJson() << std::endl;
```
Delete

sbodagala replied:
Done.

jzhou77 commented on fdbserver/ClusterController.actor.cpp:
```diff
@@ -1530,11 +1530,18 @@ ACTOR Future<Void> statusServer(FutureStream<StatusRequest> requests,
 			// requests
 			last_request_time = now();
 
+			state Optional<StatusReply> faultToleranceRelatedStatus;
```
I think you need to clear `faultToleranceRelatedStatus` after the `while` loop below. Otherwise, `faultToleranceRelatedStatus` won't be updated for the next batch.

sbodagala replied:
Yes. I missed this part.

sbodagala replied:
Addressed.

jzhou77 commented on fdbclient/include/fdbclient/StatusClient.h:
```diff
@@ -28,7 +28,7 @@
 public:
 	enum StatusLevel { MINIMAL = 0, NORMAL = 1, DETAILED = 2, JSON = 3 };
-	static Future<StatusObject> statusFetcher(Database db);
+	static Future<StatusObject> statusFetcher(Database db, std::string statusField = "");
```
Since the current implementation only supports a limited fields from the status json. It's better to document them here, so that we won't be confused in the future and assume any field can be specified.

sbodagala replied:
Yes, I will do that.

sbodagala replied:
Done.

jzhou77 commented on fdbserver/Status.actor.cpp:
```diff
@@ -3669,6 +3670,45 @@ ACTOR Future<StatusReply> clusterGetStatus(
+
+		JsonBuilderObject statusObj;
+		for (std::string& field : faultToleranceRelatedFields) {
+			if (jsonDoc.has(field, false)) {
```
is this right? The documentation of `has` says the parameter is the full path of the field.

sbodagala replied:
Yes, that's what the documentation says. I checked the code of "has()" and I think it will work even if it is not a complete path as long as this field is unique within the JSON document (here is the check it is using to locate the field: https://github.com/apple/foundationdb/blob/7295e1b806462396f0338c3c22e5f18931c5d78f/fdbclient/include/fdbclient/JSONDoc.h#L125). I can change this to be a complete path if we want to be safer about this.

jzhou77 replied:
It doesn't look right if only the "key" is used without the full path. Can you change to use the complete path? There are cases the same key is used in different paths.

sbodagala replied:
Sure.

Also, doing "setKey()" below doesn't seem to work well with how StatusClient is updating the status (https://github.com/apple/foundationdb/blob/912707f4f22932287e5ae574954e2949d3848e53/fdbclient/StatusClient.actor.cpp#L552).The code in JsonDoc/json_spirit is throwing an exception (and the error text is not very helpful in identifying the problem), looking into this issue.

sbodagala replied:
> It doesn't look right if only the "key" is used without the full path. Can you change to use the complete path?

It turns out these fields are all top-level fields in the status json object (composed by this actor). They all come under "cluster" field in the status json object output by "fdbcli", that's because "cluster" field is added by the StatusClient. So I think we can let these fields be treated as complete paths. I will drop the "false" argument here.

sbodagala replied:
> Also, doing "setKey()" below doesn't seem to work well with how StatusClient is updating the status. The code in JsonDoc/json_spirit is throwing an exception (and the error text is not very helpful in identifying the problem), looking into this issue.

I think I figured out why. We are writing the object as a string, so StatusClient is not able to look into fields of that object (that is, not able to convert it into "json_spirit::mObject::Map").

sbodagala commented:
But, are we agreeing to go ahead with this approach? If so, I will work on writing tests (I will need to check how Special keys are currently getting tested) and also I will go through the changes again looking for any finer details (and also addressing the above review comments), thanks!

jzhou77 reviewed (commented):
Yes. The general approach looks good for now. Please fix the issues and test the code.

sfc-gh-clin reviewed (commented):
Generally LGTM

johscheuer reviewed (commented):
Can we update the documentation, to make sure this new special key is documented?

sbodagala commented:
> you can add tests in SpecialKeySpaceCorrectnessWorkload or SpecialKeySpaceRobustnessWorkload

Done (added in SpecialKeySpaceCorrectnessWorkload). Please let me know if the test needs tweaking/more tests need to be added, thanks!

jzhou77 commented on fdbserver/Status.actor.cpp:
```diff
@@ -3670,6 +3671,37 @@ ACTOR Future<StatusReply> clusterGetStatus(
+		JSONDoc jsonDoc(mv);
+
+		std::string faultToleranceRelatedFields[] = {
+			"fault_tolerance", "data",    "logs", "maintenance_zone", "maintenance_seconds_remaining", "qos",
```
is this formatted correctly? I see extra spaces between "data" and "logs", and the line below.

sbodagala replied:
That's "clang-format" adding those extra spaces.

sbodagala commented:
Updated the documentation page, please take a look. Thanks!
